packages/iris-grid/src/IrisGridPartitionSelector.tsx

```diff
@@ -1,316 +1,394 @@
-  partitionTables: Table[] | null;
+  partitionConfig: PartitionConfig;
+  partitionColumnValues: readonly Items[];
+  selectorValue: readonly string[];
 }
-class IrisGridPartitionSelector<T> extends Component<
-  IrisGridPartitionSelectorProps<T>,
+class IrisGridPartitionSelector extends Component<
+  IrisGridPartitionSelectorProps,
   IrisGridPartitionSelectorState
 > {
   static defaultProps = {
     onChange: (): void => undefined,
-    onFetchAll: (): void => undefined,
-    onDone: (): void => undefined,
+    onMerge: (): void => undefined,
+    onKeyTable: (): void => undefined,
     partitions: [],
   };
 
-  constructor(props: IrisGridPartitionSelectorProps<T>) {
+  constructor(props: IrisGridPartitionSelectorProps) {
     super(props);
 
-    this.handleCloseClick = this.handleCloseClick.bind(this);
-    this.handleIgnoreClick = this.handleIgnoreClick.bind(this);
-    this.handlePartitionChange = this.handlePartitionChange.bind(this);
+    this.handleKeyTableClick = this.handleKeyTableClick.bind(this);
+    this.handleMergeClick = this.handleMergeClick.bind(this);
     this.handlePartitionSelect = this.handlePartitionSelect.bind(this);
-    this.handlePartitionListResized =
-      this.handlePartitionListResized.bind(this);
-    this.handleSearchOpened = this.handleSearchOpened.bind(this);
-    this.handleSearchClosed = this.handleSearchClosed.bind(this);
 
-    const { dh, columns, partitions } = props;
+    const { dh, columns, partitionConfig } = props;
     this.tableUtils = new TableUtils(dh);
-    this.searchMenu = columns.map(() => null);
-    this.selectorSearch = columns.map(() => null);
+    this.table = null;
+    this.partitionTables = null;
 
     this.state = {
-      partitions,
-      partitionTables: null,
+      selectorValue: columns.map(() => ''),
+      partitionConfig,
+      partitionColumnValues: columns.map(() => ({ order: [], data: {} })),
     };
   }
 
   async componentDidMount(): Promise<void> {
-    const { columns, table } = this.props;
-    const { partitions } = this.state;
+    const { columns, partitionConfig, tablePromise } = this.props;
 
-    const partitionTables = await Promise.all(
+    const table = await tablePromise;
+    this.table = table;
+    this.partitionTables = await Promise.all(
       columns.map(async (_, i) => table.selectDistinct(columns.slice(0, i + 1)))
     );
-    this.updatePartitionFilters(partitions, partitionTables);
+    this.updateConfig(partitionConfig);
+  }
+
+  componentDidUpdate(prevProps: IrisGridPartitionSelectorProps): void {
+    const { partitionConfig: prevConfig } = prevProps;
+
+    const { partitionConfig } = this.props;
+
+    if (!deepEqual(prevConfig, partitionConfig)) {
+      this.updateConfig(partitionConfig);
+    }
   }
 
   componentWillUnmount(): void {
-    const { partitionTables } = this.state;
-    partitionTables?.forEach(table => table.close());
+    this.table?.close();
+    this.partitionTables?.forEach(table => table.close());
```

Comment:
One case this _doesn't_ handle is if you unmount immediately after mount, before the tables are loading. In those cases, those tables will never get closed and we will leak them.
What you can do is make use of the `Pending` class to keep track of promises to cancel when the component unmounts. It's used in `IrisGrid` as an example, but in the constructor you'd have a `this.pending = Pending()`, and then say in `componentDidMount` where you have:
```
    this.partitionTables = await Promise.all(
      columns.map(async (_, i) => table.selectDistinct(columns.slice(0, i + 1)))
    );
```
You'd do:
```
    try {
      this.partitionTables = await Promise.all(
        // Add the inner promise to the pending queue, and add a cleanup `t => t.close()` for if it gets canceled.
        columns.map(async (_, i) => 
          pending.add(table.selectDistinct(columns.slice(0, i + 1))), t => t.close())
      );
    } catch (e) {
      if (!PromiseUtils.isCanceled(e)) {
        log.error('Unable to get partition tables', e);
      }
    }
    this.updateConfig(partitionConfig);
```

You'll need similar handling for the `awaits` in the `updateConfig` below.